Added support for multiple SQL Database CreateModes (including geo-replication) and other extension methods

LukeHarjulin opened the pull request:
**Progression with SQL Database**

Database-based geo-location works through the extension method SetAsSecondary. It sets the database to be a geo-replica of a primary db (of which the id is passed into the SetAsSecondary method)

Added more CreateMode options:
- to Restore a deleted db - use SetAsRestore()
- to "restore a geo-replication of backup" - use SetAsRecovery() (see [Link](https://www.pulumi.com/docs/reference/pkg/azure-native/sql/database/#createmode_csharp))
- to copy a db - use SetAsCopy()
- to restore a database from long term retention - use SetAsLongTermRetentionRestore()
- to restore from a point in time backup - use SetAsPointInTimeRestore()
- to set the db as a geo-replica backup (secondary) - use SetAsSecondary()

Added more ext methods:
- SecondaryType() -> not entirely sure what this is, but the default value is 'geo' (geo-replication)
- EnableZoneRedundancy() -> by default Zone Redundancy is disabled, so I made an option to enable it
- MaxDatabaseSizeGB() -> set the database size in GB with an integer (method converts to bytes and inserts into resource argument 'MaxDatabaseSizeBytes') 
- MinCapacity() -> I've not figured out what this does yet. 
- StorageAccountType -> Takes StorageAccountType struct as an argument (valid values: GRS, LRS, ZRS)
- DatabaseCollation() -> Set the collation of the database
- SampleData() -> to insert sample data into the database

Added corresponding unit test methods to test _most_ of the new extension methods (some can't be unit tested).

Most new extension methods have been tested through deployment using 'pulumi up'.

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
+        /// <param name="databaseId">Resource ID of the database to copy</param>
         /// <returns></returns>
-        public static SqlDatabaseBuilder RestoreFrom(this SqlDatabaseBuilder builder, Input<string> databaseId)
+        public static SqlDatabaseBuilder SetAsCopy(this SqlDatabaseBuilder builder, Input<string> databaseId)
```
Should we rename this to "CreateAsCopy"? I feel the current name is a bit confusing.

RichardFowles89 replied:
Why did we change "RestoreFrom" as the name? RestoreFrom.. RecoverFrom are nicely fluent.

RichardFowles89 replied:
One other thought I had - should the database ID be passed in as the argument or the database object itself and then get the ID in the method? Passing in the db object might affect the performance a little bit but the method could read "RestoreFrom(database)"  -> just a thought

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
+        /// <param name="builder"></param>
+        /// <param name="databaseId">Recoverable Database Resource ID of the database to restore</param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsRecovery(this SqlDatabaseBuilder builder, Input<string> databaseId)
```
Should we rename this to "CreateAsRecoveryOf"? I feel the current name is a bit confusing.

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
+        /// <param name="databaseId">Resource ID of the database to restore from</param>
+        /// <param name="restorePointInTime">The point in time (ISO8601 format) of the source database</param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsPointInTimeRestore(this SqlDatabaseBuilder builder, Input<string> databaseId, Input<string> restorePointInTime)
```
Should we rename this to "CreateAsPointInTimeRestore"? I feel the current name is a bit confusing.

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
+        /// <param name="builder"></param>
+        /// <param name="databaseId"></param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsLongTermRetentionRestore(this SqlDatabaseBuilder builder, Input<string> databaseId)
```
Should we rename this to "CreateAsLongTermRetentionRestore"? I feel the current name is a bit confusing.

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
+        /// <param name="deletionDate">Deletion date of the database in (ISO 8601 format).</param>
+        /// <param name="pointInTime">Point in time that you want to restore from (ISO 8601 format). Only specify to restore from an earlier point in time.</param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder SetAsRestore(this SqlDatabaseBuilder builder, Input<string> databaseId, Input<string> deletionDate, 
```
Should we rename this to "CreateAsRestoreOf"? I feel the current name is a bit confusing.

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -67,36 +66,140 @@ public static SqlDatabaseBuilder ElasticPoolId(this SqlDatabaseBuilder builder,
+        /// <param name="pointInTime">Point in time that you want to restore from. Only specify to restore from an earlier point in time.</param>
         /// <returns></returns>
-        public static SqlDatabaseBuilder SkuTier(this SqlDatabaseBuilder builder, Input<string> edition)
+        public static SqlDatabaseBuilder SetAsRestore(this SqlDatabaseBuilder builder, Input<string> databaseId, Input<string> pointInTime = null)
```
Should we rename this to "CreateAsRestoreOf"? I feel the current name is a bit confusing.

sesispla commented on src/Infrastructure/Stize.Infrastructure.Azure.Tests/Sql/Stacks/DatabaseBasicStack.cs:
```diff
@@ -23,14 +24,36 @@ public DatabaseBasicStack()
+                .SampleData(SampleName.AdventureWorksLT)
+                .Build();
+
+            var secondary = new SqlDatabaseBuilder("db1")
```
Shouldn't this second builder be named "secondaryDB"?

sesispla reviewed (changes requested):
Hi @LukeHarjulin!

Please find my feedback. I think we mainly need to make cosmetic changes to the method names.

RichardFowles89 commented on src/Infrastructure/Stize.Infrastructure.Azure/Sql/SqlDatabaseExtensions.cs:
```diff
@@ -158,5 +278,80 @@ public static SqlDatabaseBuilder SkuFamily(this SqlDatabaseBuilder builder, Inpu
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public static SqlDatabaseBuilder EnableZoneRedundant(this SqlDatabaseBuilder builder)
```
I think "EnableZoneRedundancy" might be a better name here

LukeHarjulin commented:
I've updated method names and I have added support for the user to generate a new SQL server to host the new SQL DB (to solve the second part to the [issue](https://github.com/stize/infrastructure/issues/34) you raised @sesispla ).
